streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java

```diff
@@ -0,0 +1,150 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * The {@code Branched} class is used to define the optional parameters when building branches with
+ * {@link BranchedKStream}.
+ *
+ * @param <K> type of record key
+ * @param <V> type of record value
+ */
+public class Branched<K, V> implements NamedOperation<Branched<K, V>> {
+
+    protected final String name;
+    protected final Function<? super KStream<K, V>,
+            ? extends KStream<K, V>> chainFunction;
+    protected final Consumer<? super KStream<K, V>> chainConsumer;
+
+    protected Branched(final String name,
+                       final Function<? super KStream<K, V>, ? extends KStream<K, V>> chainFunction,
+                       final Consumer<? super KStream<K, V>> chainConsumer) {
+        this.name = name;
+        this.chainFunction = chainFunction;
+        this.chainConsumer = chainConsumer;
+    }
+
+    /**
+     * Create an instance of {@link Branched} from an existing instance.
+     *
+     * @param branched the instance of {@link Branched} to copy
+     */
+    protected Branched(final Branched<K, V> branched) {
+        this(branched.name, branched.chainFunction, branched.chainConsumer);
+    }
+
+    /**
+     * Configure the instance of {@link Branched} with a branch name postfix.
+     *
+     * @param name the branch name postfix to be used. If {@code null} a default branch name postfix will be generated (see
+     *             {@link BranchedKStream} description for details)
+     * @return {@code this}
+     */
+    @Override
+    public Branched<K, V> withName(final String name) {
+        return new Branched<>(name, chainFunction, chainConsumer);
+    }
+
+    /**
+     * Create an instance of {@link Branched} with provided branch name postfix.
+     *
+     * @param name the branch name postfix to be used. If {@code null}, a default branch name postfix will be generated
+     *             (see {@link BranchedKStream} description for details)
+     * @param <K>  key type
+     * @param <V>  value type
+     * @return a new instance of {@link Branched}
+     */
+    public static <K, V> Branched<K, V> as(final String name) {
+        return new Branched<>(name, null, null);
+    }
+
+    /**
+     * Create an instance of {@link Branched} with provided chain function.
+     *
+     * @param chain A function that will be applied to the branch. If {@code null}, the identity
+     *              {@code kStream -> kStream} function will be supposed. If this function returns

```

Comment:
The KIP itself only says 'If no chain function or consumer is provided, then the value is the branch itself' which is implemented in `BranchedInternal` as following

```java
if (chainFunction == null) {
            transformedStream = newStream;
        } else {
            transformedStream = chainFunction.apply(newStream);
        }
```

So actually if `null` is passed in here, NPE won't happen, and JavaDoc just describes what actually happens. 

But you might be right! I can't think of sanely reasons to explicitly pass in `null`s , since we have an overloaded version of the static builder without `chain` param. But as a result of mistake, one can unintentionally pass `null`

Shall I throw NPEs explicitly here and in similar places, WDYT?